power_grid_model_c/power_grid_model_cpp/include/power_grid_model_cpp/dataset.hpp

```diff
@@ -0,0 +1,170 @@
+        info.handle_.check_error();
+        return n_components;
+    }
+    Idx dataset_n_components() const { return dataset_n_components(*this); }
+
+    static std::string const component_name(Info const& info, Idx component_idx) {
+        auto const component_name =
+            std::string(PGM_dataset_info_component_name(info.handle_.get(), info.info_.get(), component_idx));
+        info.handle_.check_error();
+        return component_name;
+    }
+    std::string const component_name(Idx component_idx) const { return component_name(*this, component_idx); }
+
+    static Idx component_elements_per_scenario(Info const& info, Idx component_idx) {
+        auto const elements_per_scenario =
+            PGM_dataset_info_elements_per_scenario(info.handle_.get(), info.info_.get(), component_idx);
+        info.handle_.check_error();
+        return elements_per_scenario;
+    }
+    Idx component_elements_per_scenario(Idx component_idx) const {
+        return component_elements_per_scenario(*this, component_idx);
+    }
+
+    static Idx component_total_elements(Info const& info, Idx component_idx) {
+        auto const total_elements =
+            PGM_dataset_info_total_elements(info.handle_.get(), info.info_.get(), component_idx);
+        info.handle_.check_error();
+        return total_elements;
+    }
+    Idx component_total_elements(Idx component_idx) const { return component_total_elements(*this, component_idx); }
+
+  private:
+    Handle handle_{};
+    std::unique_ptr<PGM_DatasetInfo const> info_;
+};
+
+class DatasetConst {
+  public:
+    DatasetConst(std::string const& dataset, Idx is_batch, Idx batch_size)
+        : dataset_{PGM_create_dataset_const(handle_.get(), dataset.c_str(), is_batch, batch_size)} {}
+    DatasetConst(DatasetWritable const& writable_dataset)
+        : dataset_{PGM_create_dataset_const_from_writable(handle_.get(), writable_dataset.get())} {}
+    DatasetConst(DatasetMutable const& mutable_dataset)
+        : dataset_{PGM_create_dataset_const_from_mutable(handle_.get(), mutable_dataset.get())} {}
+
+    RawConstDataset* get() const { return dataset_.get(); }
+
+    static void add_buffer(DatasetConst& dataset, std::string const& component, Idx elements_per_scenario,
+                           Idx total_elements, Idx const* indptr, RawDataConstPtr data) {
+        PGM_dataset_const_add_buffer(dataset.handle_.get(), dataset.dataset_.get(), component.c_str(),
+                                     elements_per_scenario, total_elements, indptr, data);
+        dataset.handle_.check_error();
+    }
+    void add_buffer(std::string const& component, Idx elements_per_scenario, Idx total_elements, Idx const* indptr,
+                    RawDataConstPtr data) {
+        return add_buffer(*this, component, elements_per_scenario, total_elements, indptr, data);
+    }
+
+    static Info const get_info(DatasetConst const& dataset) {
```

Comment:
returning a `const` object is dangerous, because it forces `copy`s on the users' end if they are not careful. In particular, it removes the `movable` property. Since all functions are `const` anyways, there's no harm in making it non-`const` here (and idem below)
```suggestion
    static Info get_info(DatasetConst const& dataset) {
```

Alternatively, you could make it `const&` but i don't know whether that's correct here